### PR TITLE
[no-test] WIP: webui: Introduce tmt test

### PR DESCRIPTION
**Pull request by martinpitt**
This allows us to run at least a subset of our webui integration tests in tmt, and thus run both for upstream PRs, and more importantly, as gating tests in Fedora, CentOS, and RHEL. In the future we can use this to gate packages like Firefox, LVM, systemd, and other anaconda dependencies and prevent them from breaking us.

### CODE REVIEW

File: ui/webui/test/fmf/main.fmf [hunk 1]
```diff
@@ -0,0 +1,12 @@
+summary:
+    Run webui browser integration tests on the host
+require:
+  # test dependencies
+  - chromium-headless
+  - git-core
+  - nodejs-npm
+  - libvirt-python3
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply ui/webui/test/fmf/main.fmf:8]: No, but you'll get `ImportError`s from bots if it's not installed. Perhaps we fixed that, I'll try/optimize later. I just wanted to get a first result here, but this is not ready yet.

[the remark this answers is not in the document]
Comment (martinpitt) [reply ui/webui/test/fmf/main.fmf:8]: Yep:
```
Traceback (most recent call last):
  File "/var/tmp/tmt/run-001/plans/all/discover/default-0/tests/ui/webui/test/common/run-tests", line 25, in <module>
    import testlib   # NOQA: imported through parent.py
    ^^^^^^^^^^^^^^
  File "/var/tmp/tmt/run-001/plans/all/discover/default-0/tests/ui/webui/test/common/testlib.py", line 43, in <module>
    import testvm
  File "/var/tmp/tmt/run-001/plans/all/discover/default-0/tests/ui/webui/bots/machine/testvm.py", line 31, in <module>
    from machine_core.cli import cmd_cli  # noqa: E402
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/tmp/tmt/run-001/plans/all/discover/default-0/tests/ui/webui/bots/machine/machine_core/cli.py", line 21, in <module>
    from . import machine_virtual
  File "/var/tmp/tmt/run-001/plans/all/discover/default-0/tests/ui/webui/bots/machine/machine_core/machine_virtual.py", line 29, in <module>
    import libvirt
ModuleNotFoundError: No module named 'libvirt'
```

**Comment by martinpitt** [comment]
[Bazinga](https://artifacts.dev.testing-farm.io/271476bb-3195-4f6a-a74c-bee34abf81d4/)! Now to add more tests.. And I also want to fix a few more, but that's a "tomorrow" thing.

**Comment by martinpitt** [comment]
> That said, I hope this does not gate the package once merged! That would be a disaster.

Please elaborate -- it will do that. Why don't you want that? Of course you can waive, but the whole point is to know when it breaks. Well, the bigger point is to eventually hold back *other* package updates which break anaconda (but that still needs some Fedora infra work)

But keep in mind that this will also run on every PR, so instead of learning about breakage on release, you will notice in a PR when some proposed code change breaks the gating tests.

**Comment by VladimirSlavik** [comment]
tl;dr: To gate the whole anaconda, including necessary fixes, based on an only partially functional experimental part that is not even deployed yet... makes no sense.

Anaconda is one of the critical packages for the distro(s), and the whole compose hinges on it working. That is done using the other UIs. Webui has nothing with that, it's a new experimental thing, deployed only for a subset of installations, fundamentally incapable of providing the functionality.

You just wrote that this is a gating test that allows or disallows changes - for the whole package. The WebUI build and test process is still totally bonkers, things break left and right randomly based on anything including the weather. How is this test going to be different? Are we going to have to fight this every time we need to handle something other than webui, which is everything that the ecosystem currently uses?

Actually, this does not even test what's deployed, only an approximation. Running anaconda directly on a system and pretending to install does not sound to me like a test of what we do, it's a test of what we wish to be the same.

Yes, it's a good idea, but only once webui is actually meaningfully deployed. Until then, Fedora needs anaconda gated based on what is actually used and needed, not a future wish.

BTW, it might be a good idea to ask, why do we not have any other gating tests yet? Maybe it's not so easy to actually test the right things. The system that does the gating, apart from other things, is outside of our infra... https://openqa.fedoraproject.org/

**Comment by martinpitt** [comment]
> Anaconda is one of the critical packages for the distro(s), and the whole compose hinges on it working

Fully agreed!

> The WebUI build and test process is still totally bonkers, things break left and right randomly based on anything including the weather

Why do they break so much? In many cases the way to avoid breakage is to have *more* testing, at the right places, not less of it. I.e. prevent landing changes that break things, instead of trying to catch up weeks later. You want monotonous improvement and an ever-green state, not a back-and-forth between more and less broken.

If you are landing PRs that break the web UI build, and these tests (which have been running in the Cockpit CI for a long time), then you have to stop doing that *now*. However, I don't think that actually happens -- https://github.com/rhinstaller/anaconda/pulls?q=is%3Apr+is%3Aclosed looks fairly okay; you do land some red PRs, but spot-checking a bit, one was a red coverage report and one was a red kickstart test. Landing red PRs is a high-risk thing (I didn't look into the details whether they were properly judged), but neither was about the "fedora-rawhide-boot" test, i.e. ui/webui/tests/ and its build.

> it's a new experimental thing

No, it's very much not. You even made it mission critical with https://pagure.io/fesco/issue/3032 (and also some equivalent RHEL plans). In fact so mission critical that several people from other teams got pulled in to work on it :grin: 

The web UI has been developed for over a year now, with ever-increasing effort. You can't afford to continuously break this and continuously waste time on finding regressions that happened weeks ago, that is way too expensive. Either you build a thing and *keep it working*, or you declare it a failed experiment and *stop investing* into it.  Some in-between state, like having a perpetually half-broken project, is both a waste of time and also a highly demotivating project to work on. (FTR: I don't think the web UI is in the latter state, this is just a consideration in general)

> Are we going to have to fight this every time we need to handle something other than webui, which is everything that the ecosystem currently uses?

If you propose an anaconda PR that breaks the web UI on Fedora, then yes, you have to fix it now and then, instead of "at some point later on". Of course, if something else in Fedora breaks them, i.e. they break spontaneously, that's a testing gap that means that we need to gate other Fedora updates on this test. This is exactly like the kickstart-tests, which at least used to break a lot because of Fedora changes.

-----

The above was for testing the web UI in general. *This* tmt test PR is certainly an experiment at this point. The main goal is to use this as a gating test for *other* Fedora packages to prevent beaking Anaconda. There is no reliable mechanism for that just yet, but that's what I'm trying to push forward right now (see e.g. https://packit.dev/posts/testing-farm-triggering which is also going into that direction). And for that we need tmt tests, as that's what Fedora and RHEL's gating are built on.

I asked @jkonecny12 and @KKoukiou yesterday whether it was a good idea to try and attempt this, so I started with this.  At this point, @VladimirSlavik , @jkonecny12 @KKoukiou  please let me know if you want this -- if yes, I'll spend more time in it to cover more tests. If not, I close this; no hard feelings then -- it was a fun thing to do, it already identified some code which needs to be cleaned up, and it didn't waste too much time.

**Comment by rvykydal** [comment]
> ... Of course, if something else in Fedora breaks them, i.e. they break spontaneously, that's a testing gap that means that we need to gate other Fedora updates on this test. This is exactly like the kickstart-tests, which at least used to break a lot because of Fedora changes.

Yes, I think this has been my biggest concern (apart from technical issues) preventing using kickstart-tests for gating (gating of package updates, gating of PRs is done with just smoke tests which are far less sensitive to environment changes then the whole kstest suite or its COVERAGE subset). But having these tests actually gating changes in our dependencies would be great and something we want IMO. Then we should also deploy some kickstart tests for gating.

**Comment by rvykydal** [comment]
> Actually, this does not even test what's deployed, only an approximation. Running anaconda directly on a system and pretending to install does not sound to me like a test of what we do, it's a test of what we wish to be the same.

If I understand it correctly tmt is not much further from the real installer environment then our current CI webui tests are right? But still it is perhaps yet another environment to keep the tests working in ...

**Comment by martinpitt** [comment]
> If I understand it correctly tmt is not much further from the real installer environment then our current CI webui tests are right?

It is quite a bit different. Our current tests actually boot from `boot.iso` and install into the given disk image, whereas a TF instance cannot (currently) do this -- boot.iso isn't something you can boot on a cloud, it would at least need to grow a way of getting unattended SSH access. Specifically to the TF, I at least don't know a way how to select a custom image, although I've heard that this has been on the list for a while.

So this draft just boots a standard Rawhide cloud image, installs the anaconda RPMs and runs it from there. It is primarily meant to exercise the user interface, and thus make sure that things like Firefox, blivet, systemd, etc. keep working. I do want to extend it to actually do a (mini) install into a new disk (TF can attach an extra disk, for example; or an LV-on-loop, which is easier for local reproduction). But I'm putting a hold on this until I know that you actually want to do this.

**Comment by martinpitt** [comment]
We discussed that interactively: Summary from my POV:
 - Let's at least stall this until we have an actual mechanism in Fedora and/or upstream projects to do actual reverse dependency testing (I have a meeting about that on Monday)
 - Concern was that the environment is too different from boot.iso to be representative
 - Getting the storage tests running in tmt could be useful
 - There was no recent breakage of Fedora package updates that broke the web UI, other than a Firefox incident which only happened in the live environment
 - It would be more useful to get the backend tests running in tmt (that's also a good idea, but not an either-or -- you can add many test plans/tests)

**Comment by martinpitt** [comment]
Putting this on the shelf. Let me know if/when you are interested in this, then I can spend some more time on that. I have no direct personal pressure for this.
